### PR TITLE
Move `PostgresSourceReader` to `SourceReader` over `SimpleSource`

### PR DESCRIPTION
**Pull request by guswynn**
Ultimately, I think we should not have multiple `Source` traits. This is the first such change for this: switching `Postgres` off of a `SimpleSource` to a full `SourceReader`. Note that this gives us timestamp bindings for pg for free!

This pr has several subtle parts:
- The first commit resolves an outstanding bug: `SourceReader` assumes that a message at some offset the last such message. Pg has transactions that happen at the same timestamp, so we need to support holding back a cursor to support this. This is quite subtle, and @petrosagg may be interested in looking closely
- The second commit adds support for sources that _retract_ (that is, dd diffs that ARENT `1`) values. This is to support both replication failure (causing a revert), and pg deletes. 
	- This pr uses new associated types to avoid unwraps as much as possible, but to keep `render_source` readable, forces a `diff` in the value in `DecodeResult`.
	-  We add `unreachable!` for not-possible `diff` values in envelopes that don't support esoteric diffs (namely, things other than `None`. Its not 100% clear to me if `KeyEnvelope::None` is the only supported `SourceEnvelope::None`, but the coden as-written does not deal with 
- The third commit actually moves pg to a `SourceReader`. As much code as possible is preserved, and I opted for the `get_next_message` route over `next`, as we can transform all sources at the same time
- The final commit ensures that pg sources are not considered monotonic. This appears to have already been an outstanding bug

### Motivation

  * This PR adds a known-desirable feature.

https://github.com/MaterializeInc/materialize/issues/11899

### Tips for reviewer

I recommend reviewing the commits in this pr separately. 

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

None

### CODE REVIEW

File: src/dataflow-types/src/types.rs [hunk 1]
```diff
@@ -769,12 +769,12 @@ pub mod sources {
     #[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
     pub enum SourceEnvelope {
         /// The most trivial version is `None`, which typically produces triples where the diff
-        /// is ALWAYS `1`
+        /// is `1`. However, some sources are able to produce values with more exotic diff's,
+        /// such as the posgres source. Currently, this is the only variant usable with
+        /// those sources.
```
Comment (frankmcsherry) [inline src/dataflow-types/src/types.rs:774]: Can you say a bit about why this change? Up until now `Envelope::None` had meant "append", and .. it now means something else.

Comment (guswynn) [reply src/dataflow-types/src/types.rs:774]: This was actually already broken: `SimpleSource`'s could already retract/delete rows, and they used `SourceEnvelope::None`

Comment (frankmcsherry) [reply src/dataflow-types/src/types.rs:774]: Pushing the bug further down seems like maybe the wrong outcome?

Comment (guswynn) [reply src/dataflow-types/src/types.rs:774]: @frankmcsherry I am not sure it is pushing it down, I fixed the `optimize_dataflow_monotonic`, which as far as I can tell, was the only erroneous assumption. that said, I can definitely add an envelope and clean up the sql parsing and type-checking in a later pr!

Comment (frankmcsherry) [reply src/dataflow-types/src/types.rs:774]: Let me restate: historically `None` for external sources meant "append only" and was confirmed by the sources/sinks folks to mean that. I agree it was never documented. Asking `transform` (not a storage-y crate) to track which combinations of envelopes are "append only" seems like an abstraction violation. STORAGE should provide a `fn is_append_only` method on some sort of source identifier, so that the `transform` crate can rely on it rather than the shifting meanings of this enum.

Comment (guswynn) [reply src/dataflow-types/src/types.rs:774]: @frankmcsherry Oh see! Yes I will change it to that!

Comment (petrosagg) [reply src/dataflow-types/src/types.rs:763]: @guswynn @aljoscha has added a new Envelope type called `DifferentialRow` in order to discern between append-only and collection-like things. Should use that for postgres sources?

Comment (guswynn) [reply src/dataflow-types/src/types.rs:763]: @petrosagg I think so, pg cant have a `KeyEnvelope` other than `None` right?

Comment (guswynn) [reply src/dataflow-types/src/types.rs:1046]: https://github.com/MaterializeInc/materialize/issues/12413

File: src/storage/src/source/mod.rs [hunk 2]
```diff
@@ -986,13 +1041,31 @@ where
...
+                                // 1. Requires that sources that produce `InProgress` messages
+                                //    ALWAYS produce a `Finalized` for the final message.
+                                // 2. Requires that sources that produce `InProgress` messages
+                                //    NEVER produces messages at offsets below the most recent
+                                //    `Finalized` message.
+                                // 3. Buffers EVERY message associated with a single offset. This
+                                //    can be improved, tracked in
+                                //    <https://github.com/MaterializeInc/materialize/issues/12557>
+                                SourceMessageType::Finalized(message) => {
+                                    pending_messages.extend(pending_untimestamped_messages.drain(..));
```
Comment (cjubb39) [inline src/storage/src/source/mod.rs:1062]: should this be keyed by partition id?  Offset is a property of the partition (perhaps this doesn't matter at the moment because the set of partitioned sources and the set of sources that will return `InProgress` may be non-overalpping

File: src/storage/src/render/debezium.rs [hunk 3]
```diff
@@ -62,7 +62,10 @@ where
                                 }
                             };
                             let value = match result.value {
-                                Some(Ok(value)) => value,
+                                Some(Ok((value, 1))) => value,
+                                Some(Ok(_)) => unreachable!(
```
Comment (cjubb39) [inline src/storage/src/render/debezium.rs:66]: where is this checked?

File: src/storage/src/render/sources.rs [hunk 4]
```diff
@@ -631,7 +635,7 @@ where
 #[derive(Debug, Clone, Hash, PartialEq, Eq, Ord, PartialOrd, Serialize, Deserialize)]
 struct KV {
     key: Option<Result<Row, DecodeError>>,
-    val: Option<Result<Row, DecodeError>>,
+    val: Option<Result<(Row, i64), DecodeError>>,
```
Comment (cjubb39) [inline src/storage/src/render/sources.rs:638]: nit: `i64` -> `Diff`? To make it more clear what the integer represents?

File: src/storage/src/source/mod.rs [hunk 5]
```diff
@@ -478,9 +527,15 @@ pub struct SourceMessage<Key, Value> {
     /// Headers, if the source is configured to pass them along. If it is, but there are none, it
     /// passes `Some([])`
     pub headers: Option<Vec<(String, Option<Vec<u8>>)>>,
+
+    /// Allow sources to optionally output a specific differential
+    /// `diff` value. Defaults to `+1`.
```
Comment (cjubb39) [inline src/storage/src/source/mod.rs:532]: I'm curious why you've structured it as an arbitrary `Diff` type rather than either:
- Option<mz_repr::Diff>
- `mz_repr::Diff` and having everything other than postgres set it to 1 (since they can authoritatively say that)

I don't follow (perhaps out of ignorance!) when we'd need to support a non-`mz_repr::Diff` in this field?

File: src/storage/src/source/postgres.rs [hunk 6]
```diff
@@ -130,28 +128,292 @@ macro_rules! try_recoverable {
...
-impl PostgresSourceReader {
-    /// Constructs a new instance
-    pub fn new(
+// Message used to communicate between `get_next_message` and the tokio task
+enum InternalMessage {
+    Err(SourceReaderError),
+    Value {
+        value: Row,
+        lsn: PgLsn,
+        diff: i64,
```
Comment (cjubb39) [inline src/storage/src/source/postgres.rs:137]: same nit here: `i64` -> `mz_repr::Diff`?

File: src/storage/src/source/postgres.rs [hunk 7]
```diff
@@ -130,28 +128,292 @@ macro_rules! try_recoverable {
...
+        connector: ExternalSourceConnector,
+        _restored_offsets: Vec<(PartitionId, Option<MzOffset>)>,
+        _encoding: SourceDataEncoding,
+        metrics: SourceBaseMetrics,
+        _connector_context: ConnectorContext,
+    ) -> Result<Self, anyhow::Error> {
+        let connector = match connector {
+            ExternalSourceConnector::Postgres(pg) => pg,
+            _ => {
+                panic!("S3 is the only legitimate ExternalSourceConnector for S3SourceReader")
```
Comment (cjubb39) [inline src/storage/src/source/postgres.rs:180]: where is this checked?

**Review by cjubb39 — COMMENTED** [review]
gave a quick look-through (didn't get to the actual logic of the postgres source reader impl  so it's possible that some suggestions are redundant.  In that case sorry!)

File: src/storage/src/source/mod.rs (hunk 2 above)
Comment (guswynn) [reply src/storage/src/source/mod.rs:1062]: oh yes, it should be a map with a partition key!

File: src/storage/src/render/debezium.rs (hunk 3 above)
Comment (guswynn) [reply src/storage/src/render/debezium.rs:66]: in typechecking, we don't allow `ENVELOPE DEBEZIUM` with pg sources

File: src/storage/src/source/mod.rs (hunk 5 above)
Comment (guswynn) [reply src/storage/src/source/mod.rs:532]: I had it `Option<mz_repr::Diff>` or something similar but I found that it made it unclear in `render_source` what was happening. with the associated type we can ensure that ONLY
`SourceType::Row` ends up having `type Diff = Diff`, where as other types always have `type Diff = ()` .

in the future this also enables us to generic-ize `DecodeResult` and cleanup `render_source` even more

File: src/storage/src/source/postgres.rs (hunk 7 above)
Comment (guswynn) [reply src/storage/src/source/postgres.rs:180]: ugh, this is the wrong message, but this is the same check that every source does!

File: src/storage/src/source/mod.rs (hunk 5 above)
Comment (guswynn) [reply src/storage/src/source/mod.rs:532]: so in short: yes, we want other diff types, namely `()` to clearly communicate when a source wants to use the default behavior, as opposed to all those sources needing to implement that default themselves

File: src/storage/src/source/postgres.rs [hunk 8]
```diff
@@ -370,272 +627,211 @@ impl PostgresSourceReader {
...
+                                            .collect::<Vec<String>>();
                                         return Err(Fatal(anyhow!(
-                                            "source table {} with oid {} has been altered",
-                                            source_table.name,
-                                            source_table.relation_id
+                                            "source table(s) {} got truncated",
+                                            tables.join(", ")
                                         )));
                                     }
+                                    // The enum is marked as non_exaustive. Better to be conservative here in
```
Comment (cjubb39) [inline src/storage/src/source/postgres.rs:820]: `non_exhaustive`

File: src/storage/src/source/postgres.rs [hunk 9]
```diff
@@ -370,272 +627,211 @@ impl PostgresSourceReader {
...
+                                    // case a new message is relevant to the semantics of our source
+                                    _ => return Err(Fatal(anyhow!("unexpected logical replication message"))),
                                 }
-                                // Ignore messages for tables we do not know about
-                                None => continue,
                             }
+                            // Handled above
+                            PrimaryKeepAlive(_) => {}
+                            // The enum is marked non_exaustive, better be conservative
```
Comment (cjubb39) [inline src/storage/src/source/postgres.rs:827]: `non_exhaustive`

File: src/storage/src/source/postgres.rs [hunk 10]
```diff
@@ -370,272 +627,211 @@ impl PostgresSourceReader {
...
-                            let old_tuple = try_fatal!(delete
-                                .old_tuple()
-                                .ok_or_else(|| anyhow!("Old row missing from replication stream for table with OID = {}. \
-                                        Did you forget to set REPLICA IDENTITY to FULL for your table?", rel_id)))
-                            .tuple_data();
-                            let row = try_fatal!(self.row_from_tuple(rel_id, old_tuple));
-                            deletes.push(row);
+        let closer = self.sender.clone();
+        loop {
+            tokio::select! {
```
Comment (cjubb39) [inline src/storage/src/source/postgres.rs:632]: mind adding a comment like "yo these are both cancel-safe"? (so long as that fits into the grander plan for `tokio::select`)

File: src/dataflow-types/src/types.rs [hunk 11]
```diff
@@ -2173,6 +2173,31 @@ pub mod sources {
             }
         }
 
+        /// Returns `true` if this connector yields data that is
+        /// append-only/monotonic. Append-monly means the source
+        /// never produces retractions.
+        // TODO(guswynn): consider enforcing this more completely at the
+        // parsing/typechecking level, by not using an `envelope`
+        // for sources like pg
+        pub fn append_only(&self) -> bool {
```
Comment (cjubb39) [inline src/dataflow-types/src/types.rs:2182]: I know it's more typing but what do you think of being more explicit with our types in the match here? In particular I think it would provide value to force anyone who adds a `SourceEnvelope` / `SourceConnector` / `ExternalSourceConnector` variant to think about whether it is in fact append only

File: src/storage/src/source/mod.rs (hunk 5 above)
Comment (guswynn) [reply src/storage/src/source/mod.rs:532]: with a fixed type, either `diff: i64` or `diff: Option<i64>`, the `SourceReader` impl is required to provide the "default" to us. If kafka's `SourceReader` gives us `2` over `None` or `1`, we can't really tell if thats wrong or not, we have to assume its right

With this, we declare, at the top of the trait impl, that the kafka source reader can produce only _1_ value as a diff (in this case `()`, but we could make an empty struct call `Default` or whatever, preventing any complexity or thought in the decision when producing the `SourceMessage` inside the `next` method.

I found this to be much cleaner and easier to reason about, and less fragile. You are right that it doesn't buy us much, but I am inclined to encode invariants in the type system as much as possible, without making stuff unreadable. I found this to be more readable than the previous version (you may be able to find a commit on this pr from earlier) that didn't do it this way

File: src/storage/src/source/mod.rs [hunk 12]
```diff
@@ -994,13 +1049,47 @@ where
...
+                                }
+                                SourceMessageType::InProgress(message) => {
+                                    // this extra if-statement is just here to avoid a clone in
+                                    // case we have expensive partition id's someday
+                                    if let Some(untimestamped_messages) = untimestamped_messages.get_mut(
+                                        &message.partition
+                                    ) {
+                                        untimestamped_messages.push(Ok(message))
+                                    } else {
+                                        untimestamped_messages
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/mod.rs:1086]: oh I just was trying to clone the pid one more time, but im probably over-optimizing

File: src/storage/src/source/postgres.rs [hunk 13]
```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
...
+        _connector_context: ConnectorContext,
+    ) -> Result<Self, anyhow::Error> {
+        let connector = match connector {
+            ExternalSourceConnector::Postgres(pg) => pg,
+            _ => {
+                panic!("Postgres is the only legitimate ExternalSourceConnector for PostgresSourceReader")
+            }
+        };
+
+        let (dataflow_tx, dataflow_rx) = tokio::sync::mpsc::channel(10_000);
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs:182]: yes, it will just put the pg task to sleep till it has space (but a later pr will probably remove the task and move it to the `SourceReader::next` api

File: src/storage/src/source/postgres.rs [hunk 14]
```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
...
+            receiver_stream: dataflow_rx,
+        })
+    }
+
+    // TODO(guswynn): use `next` instead of using a channel
+    fn get_next_message(
+        &mut self,
+    ) -> Result<NextMessage<Self::Key, Self::Value, Self::Diff>, SourceReaderError> {
+        // TODO(guswynn): consider if `try_recv` is better or the same as `now_or_never`
+        let ret = match self.receiver_stream.recv().now_or_never() {
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs:215]: yeah, I wanted to keep this pr as original as possible, and switch to the `SourceReader::next` api in a later pr!

File: src/storage/src/source/postgres.rs [hunk 15]
```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
...
+            Some(Some(InternalMessage::Value {
+                value,
+                diff,
+                lsn,
+                end,
+            })) => {
+                let offset: u64 = lsn.into();
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs:222]: I think I didnt do that just to avoid adding the pg dep in `mz-dataflow-types`? if you think that is okay, then I can do that! I will eventually switch offset to u64 and avoid any unwraps at all

File: src/storage/src/source/postgres.rs [hunk 16]
```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
...
+async fn postgres_replication_loop_inner(
+    task_info: &mut PostgresTaskInfo,
+) -> Result<(), SourceReaderError> {
+    let source_id = task_info.source_id.clone();
+    // Buffer rows from snapshot to retract and retry, if initial snapshot fails.
+    // Postgres sources cannot proceed without a successful snapshot.
+    let mut snapshot_tx = Transaction::new();
+    loop {
+        let file =
+            tokio::fs::File::from_std(tempfile::tempfile().map_err(|e| SourceReaderError {
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs:288]: this is just for the revert path, if the snapshot fails to complete, but we def have it on the backburner to fix this for platform!

File: src/storage/src/source/postgres.rs [hunk 17]
```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
...
+    }
+
+    /// Record a deletion of a row in the current transaction
+    pub fn delete(&mut self, row: Row) {
+        self.rows.push((row, -1));
+    }
+
+    /// Finalize a transaction and send it off through the channel
+    ///
+    /// Care must be taken to ALWAYS call this if inserts/deletes have occurred
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs:384]: I don't think so, because error's can cause early-returns, where we don't care what happens

File: src/storage/src/source/postgres.rs [hunk 18]
```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
...
+                    value: row,
+                    lsn,
+                    diff,
+                    end: i == (num - 1),
+                })
+                .await
+            {
+                activator
+                    .activate()
+                    .expect("postgres reader activation failed");
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs:407]: this one I am unsure the origin of, I think I just copied the s3 code?

File: src/dataflow-types/src/types.rs (hunk 11 above)
Comment (guswynn) [reply src/dataflow-types/src/types.rs:2182]: great point!

File: src/storage/src/source/postgres.rs [hunk 19]
```diff
@@ -128,28 +126,292 @@ macro_rules! try_recoverable {
...
+    match postgres_replication_loop_inner(&mut task_info).await {
+        Ok(()) => {}
+        Err(e) => {
+            // Drop the send error, as we have no way of communicating back to the
+            // source operator if the channel is gone.
+            let _ = task_info.sender.send(InternalMessage::Err(e)).await;
+            task_info
+                .activator
+                .activate()
+                .expect("postgres reader activation failed");
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/source/postgres.rs:273]: cc @petrosagg

File: src/storage/src/source/postgres.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (cjubb39) [reply src/storage/src/source/postgres.rs:222]: I don't see why not!  imo it's a reasonable dependency!  That crate depends on `kafka-util`, which, in turn, depends on `rdkafka`

File: src/storage/src/source/postgres.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (cjubb39) [reply src/storage/src/source/postgres.rs:288]: can you explain more? it looks like the buffer created from the file is used to hold data inside `produce_snapshot`?

[the remark this answers is not in the document]
Comment (cjubb39) [reply src/storage/src/source/postgres.rs:288]: actually -- don't worry about it.  it's not particularly relevant to your PR and sounds like it's on the docket to fix so no worries

File: src/storage/src/source/postgres.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (cjubb39) [reply src/storage/src/source/postgres.rs:384]: ahh true that does make sense -- in that case, mind updating the comment since it isn't quite correct?

File: src/storage/src/source/postgres.rs (hunk 10 above)
Comment (guswynn) [reply src/storage/src/source/postgres.rs:632]: oh interesting its more subtle than that...but still safe I thini...luckily we can just move to `SourceReader::next` in the future and avoid this altogether

File: src/storage/src/source/postgres.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (cjubb39) [reply src/storage/src/source/postgres.rs:407]: 🤷‍♀️ okay sure -- that's good enough for me!

File: src/storage/src/source/mod.rs (hunk 5 above)
Comment (cjubb39) [reply src/storage/src/source/mod.rs:532]: > If kafka's SourceReader gives us 2 over None or 1, we can't really tell if thats wrong or not, we have to assume its right

Isn't `KafkaSourceReader` the authoritative source for what is correct for a kakfa source though? Just like the `PostgresSourceReader` is the authoritative source for the diff field for a postgres source.  We're not supposing what we should protect against postgres being wrong.  I think that's the thing I'm not understanding.

Hmm I guess this setup is analogous to parametrizing over a trait implemented by two enums: `enum AppendOnly { Append }` and `enum AppendOrDelete { Append, Delete }` -- which I do very clearly see the value in soperhaps it's the naming tripping me up? I think writing those enums out is more explicit (and maybe even more clear but I was the one who came up with it so grain of salt...) but if you don't want to do that, spitballing here: what about just renaming `specific_diff -> update_type`?

File: src/dataflow-types/src/types.rs (hunk 11 above)
Comment (guswynn) [reply src/dataflow-types/src/types.rs:2182]: `or_patterns` stabilizing makes this much cleaner

File: src/storage/src/source/mod.rs (hunk 5 above)
Comment (guswynn) [reply src/storage/src/source/mod.rs:532]: @cjubb39 `KafkaSourceReader` isnt really the authoritative source: it can be composed with envelopes that may or may not support non-`1` diffs; so in some sense, the caller in `render::sources::render_source` is authoritative (as it should be, it is the place where `SourceReaders` are correctly instantiated

we can definitely change the name! `update_type` is fine, what about `type MessageDiff` and `message_diff`?

**Comment by guswynn** [comment]
@cjubb39 I believe I resolved all your comments!

**Review by cjubb39 — APPROVED** [review]
please update the comment on the MzOffset impl below before merging! will leave the other one up to you

**Comment by cjubb39** [comment]
Thanks for doing this change! Happy to moving to have one impl for sources!
